src/xrpld/app/misc/LendingHelpers.h

```diff
@@ -0,0 +1,258 @@
+    Number interestDue;
+    /// Management fee still due to be paid TO the broker.
+    // This is a portion of interestOutstanding
+    Number managementFeeDue;
+
+    /// Interest still due to be paid by the borrower.
+    Number
+    interestOutstanding() const
+    {
+        XRPL_ASSERT_PARTS(
+            interestDue + managementFeeDue ==
+                valueOutstanding - principalOutstanding,
+            "ripple::LoanState::interestOutstanding",
+            "other values add up correctly");
+        return interestDue + managementFeeDue;
+    }
+};
+
+LoanState
+calculateRawLoanState(
+    Number const& periodicPayment,
+    Number const& periodicRate,
+    std::uint32_t const paymentRemaining,
+    TenthBips32 const managementFeeRate);
+
+LoanState
+calculateRawLoanState(
+    Number const& periodicPayment,
+    TenthBips32 interestRate,
+    std::uint32_t paymentInterval,
+    std::uint32_t const paymentRemaining,
+    TenthBips32 const managementFeeRate);
+
+LoanState
+calculateRoundedLoanState(
+    Number const& totalValueOutstanding,
+    Number const& principalOutstanding,
+    Number const& managementFeeOutstanding);
+
+LoanState
+calculateRoundedLoanState(SLE::const_ref loan);
+
+Number
+computeFee(
+    Asset const& asset,
+    Number const& value,
+    TenthBips32 managementFeeRate,
+    std::int32_t scale);
+
+Number
+calculateFullPaymentInterest(
+    Number const& rawPrincipalOutstanding,
+    Number const& periodicRate,
+    NetClock::time_point parentCloseTime,
+    std::uint32_t paymentInterval,
+    std::uint32_t prevPaymentDate,
+    std::uint32_t startDate,
+    TenthBips32 closeInterestRate);
+
+Number
+calculateFullPaymentInterest(
+    Number const& periodicPayment,
+    Number const& periodicRate,
+    std::uint32_t paymentRemaining,
+    NetClock::time_point parentCloseTime,
+    std::uint32_t paymentInterval,
+    std::uint32_t prevPaymentDate,
+    std::uint32_t startDate,
+    TenthBips32 closeInterestRate);
+
+namespace detail {
+// These classes and functions should only be accessed by LendingHelper
+// functions and unit tests
+
+enum class PaymentSpecialCase { none, final, extra };
+
+/// This structure is used internally to compute the breakdown of a
+/// single loan payment
+struct PaymentComponents
+{
+    // tracked values are rounded to the asset and loan scale, and correspond to
+    // fields in the Loan ledger object.
+    // trackedValueDelta modifies sfTotalValueOutstanding.
+    Number trackedValueDelta;
+    // trackedPrincipalDelta modifies sfPrincipalOutstanding.
+    Number trackedPrincipalDelta;
+    // trackedManagementFeeDelta modifies sfManagementFeeOutstanding. It will
+    // not include any "extra" fees that go directly to the broker, such as late
+    // fees.
+    Number trackedManagementFeeDelta;
+
+    PaymentSpecialCase specialCase = PaymentSpecialCase::none;
+
+    Number
+    trackedInterestPart() const;
+};
+
+struct LoanDeltas
+{
+    Number principalDelta;
```

Comment:
Probably don't need to have `Delta` in the name since the structure name includes `Deltas`.

Reply:
> Probably don't need to have `Delta` in the name since the structure name includes `Deltas`.

These fields have since been renamed.